Mathlib/AlgebraicGeometry/Morphisms/Basic.lean

```diff
@@ -156,347 +233,248 @@ instance targetAffineLocally_respectsIso (P : AffineTargetMorphismProperty)
+  isLocal_affineProperty := e' ▸ isLocal_affineProperty P
+  eq_targetAffineLocally' := e' ▸ e.symm ▸ eq_targetAffineLocally P
+
+variable {P}
+
+theorem of_isPullback {U V} [IsAffine U] {g : U ⟶ Y} [IsOpenImmersion g]
+    {iV : V ⟶ X} {f' : V ⟶ U} (h : IsPullback iV f' f g) (hf : P f) :
+    Q f' :=
+  letI := isLocal_affineProperty P
+  of_targetAffineLocally_of_isPullback h (eq_targetAffineLocally (P := P) ▸ hf)
+
+theorem restrict (hf : P f) (U : Y.affineOpens) :
+    Q (f ∣_ U) :=
+  of_isPullback (isPullback_morphismRestrict f U).flip hf
+
+instance (priority := 900) : P.RespectsIso := by
+  letI := isLocal_affineProperty P
+  rw [eq_targetAffineLocally P]
+  infer_instance
+#align algebraic_geometry.target_affine_locally_respects_iso AlgebraicGeometry.HasAffineProperty.instRespectsIsoScheme
+
+theorem of_iSup_eq_top
+    {ι} (U : ι → Y.affineOpens) (hU : ⨆ i, (U i : Opens Y) = ⊤)
+    (hU' : ∀ i, Q (f ∣_ U i)) :
+    P f := by
+  letI := isLocal_affineProperty P
+  rw [eq_targetAffineLocally P]
+  classical
+  intro V
+  induction V using of_affine_open_cover U hU  with
+  | basicOpen U r h =>
+    haveI : IsAffine _ := U.2
+    have := AffineTargetMorphismProperty.IsLocal.to_basicOpen (f ∣_ U.1) ((Y.resTop U).inv r) h
+    exact (Q.arrow_mk_iso_iff
+      (morphismRestrictRestrictBasicOpen f _ r)).mp this
+  | openCover U s hs H =>
+    apply AffineTargetMorphismProperty.IsLocal.of_basicOpenCover _
+      (s.image (Y.resTop _).inv) (by simp [← Ideal.map_span, hs, Ideal.map_top])
+    intro ⟨r, hr⟩
+    obtain ⟨r, hr', rfl⟩ := Finset.mem_image.mp hr
+    exact (Q.arrow_mk_iso_iff
+      (morphismRestrictRestrictBasicOpen f _ r).symm).mp (H ⟨r, hr'⟩)
+  | hU i => exact hU' i
+
+theorem iff_of_iSup_eq_top
+    {ι} (U : ι → Y.affineOpens) (hU : ⨆ i, (U i : Opens Y) = ⊤) :
+    P f ↔ ∀ i, Q (f ∣_ U i) :=
+  ⟨fun H _ ↦ restrict H _, fun H ↦ HasAffineProperty.of_iSup_eq_top U hU H⟩
+
+theorem of_openCover
+    (𝒰 : Y.OpenCover) [∀ i, IsAffine (𝒰.obj i)] (h𝒰 : ∀ i, Q (𝒰.pullbackHom f i)) :
+    P f :=
+  letI := isLocal_affineProperty P
+  of_iSup_eq_top
+    (fun i ↦ ⟨_, isAffineOpen_opensRange (𝒰.map i)⟩) 𝒰.iSup_opensRange
+    (fun i ↦ (Q.arrow_mk_iso_iff (morphismRestrictOpensRange f _)).mpr (h𝒰 i))
+#align algebraic_geometry.target_affine_locally_of_open_cover AlgebraicGeometry.HasAffineProperty.of_openCover
+
+theorem iff_of_openCover (𝒰 : Y.OpenCover) [∀ i, IsAffine (𝒰.obj i)] :
+    P f ↔ ∀ i, Q (𝒰.pullbackHom f i) := by
+  letI := isLocal_affineProperty P
+  rw [iff_of_iSup_eq_top (P := P)
+    (fun i ↦ ⟨_, isAffineOpen_opensRange _⟩) 𝒰.iSup_opensRange]
+  exact forall_congr' fun i ↦ Q.arrow_mk_iso_iff
+    (morphismRestrictOpensRange f _)
+
+theorem morphsimRestrict (U : Opens Y) (H : P f) : P (f ∣_ U) := by
+  letI := isLocal_affineProperty P
+  rw [eq_targetAffineLocally P] at H ⊢
+  intros V
+  rw [Q.arrow_mk_iso_iff (morphismRestrictRestrict _ _ _)]
+  exact H ⟨_, V.2.image_of_isOpenImmersion _⟩
+
```

Comment:
I don't think this lemma is necessary, as it can be obtained from `IsLocalAtTarget.restrict`.